framework/src/base/Assembly.C

```diff
@@ -944,14 +944,15 @@ Assembly::computeAffineMapAD(const Elem * elem,
         unsigned dimension = 0;
         if (_computing_jacobian)
           for (const auto & disp_num : _displacements)
-            Moose::derivInsert(elem_point(dimension++).derivatives(),
+            if (node.n_dofs(sys_num, disp_num))
```

Comment:
As far as I understand this was possibly doing some unnecessary work before this PR?

Reply:
More insidious than that...this would put derivatives into these geometric quantities even if the displacement variables were auxiliary variables. So you can imagine that if the aux var displacements were numbered 0 and 1 in the aux system, and you had some nl variables, let's say T and u numbered 0 and 1 respectively in the nl system, you would actually be inserting geometric quantity dependence on T and u!

Reply:
Wait this check is actually still bad... we're still going to enter this block under the exact scenario I posted above anytime T and u have dofs on the node!

Reply:
Whew, actually things are ok because of this code:
```
void
DisplacedProblem::init()
{
  for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
  {
    _assembly[tid]->init(_mproblem.couplingMatrix());
    std::vector<unsigned> disp_numbers;
    for (const auto & disp_string : _displacements)
    {
      const auto & disp_variable = getVariable(tid, disp_string);
      if (disp_variable.kind() == Moose::VarKindType::VAR_NONLINEAR)
        disp_numbers.push_back(disp_variable.number());
    }
    _assembly[tid]->assignDisplacements(std::move(disp_numbers));
  }
```
I check to make sure the displacement variable is nonlinear before passing it to `Assembly`.

Ok so after that whirlwind: the real reason this check is necessary is in case you are at a place in the mesh where the displacement variable doesn't actually exist. If we don't do this `n_dofs` check here, then we would fail an assertion when we call `node.dof_number`